File: clj/src/day06/guard_map_test.clj
(ns day06.guard-map-test
  (:require [clojure.test :refer [deftest is testing]]
            [day06.guard-map :as gm]))

(deftest day06
  (testing "parse-safety-manual-updates-file"
    (testing "reads pages orders header {before}|{after}"
      (is (= {:map ["....#....."
                    ".........#"
                    ".........."
                    "..#......."
                    ".......#.."
                    ".........."
                    ".#..^....."
                    "........#."
                    "#........."
                    "......#..."]
              :width 10
              :height 10}
             (gm/parse-guard-map-file {:file-name "data/day06/test.txt"})))))
  (testing "->initial-guard-position"
    (testing "extract initial guard position marked by '^'"
      (let [guard-map (gm/parse-guard-map-file {:file-name "data/day06/test.txt"})]
        (is (= {:direction :up :row 6 :col 4}
               (gm/->initial-guard-position {:guard-map guard-map}))))))
  (testing "guard-exiting?"
    (testing "checks if advancing guard will exit the map"
      (is (= false
             (gm/guard-exiting? {:guard-map {:height 10 :width 10}
                                 :guard-position {:direction :up :row 6 :col 4}}))
          "starting")
      (is (= true
             (gm/guard-exiting? {:guard-map {:height 10 :width 10}
                                 :guard-position {:direction :up :row 0 :col 4}}))
          "up")
      (is (= true
             (gm/guard-exiting? {:guard-map {:height 10 :width 10}
                                 :guard-position {:direction :down :row 9 :col 4}}))
          "down")
      (is (= true
             (gm/guard-exiting? {:guard-map {:height 10 :width 10}
                                 :guard-position {:direction :left :row 6 :col 0}}))
          "left")
      (is (= true
             (gm/guard-exiting? {:guard-map {:height 10 :width 10}
                                 :guard-position {:direction :right :row 6 :col 9}}))
          "left")))
  (testing "guard-next-position"
    (testing "advance guard position depending on their direction"
      (is (= {:direction :up :row 5 :col 4}
             (gm/guard-next-position {:guard-position {:direction :up :row 6 :col 4}})))
      (is (= {:direction :down :row 7 :col 4}
             (gm/guard-next-position {:guard-position {:direction :down :row 6 :col 4}})))
      (is (= {:direction :left :row 6 :col 3}
             (gm/guard-next-position {:guard-position {:direction :left :row 6 :col 4}})))
      (is (= {:direction :right :row 6 :col 5}
             (gm/guard-next-position {:guard-position {:direction :right :row 6 :col 4}})))))
  (testing "turn-guard-right"
    (testing "changes guard direction to the right"
      (is (= {:direction :up :row 6 :col 4}
             (gm/turn-guard-right {:guard-position {:direction :left :row 6 :col 4}})))
      (is (= {:direction :right :row 6 :col 4}
             (gm/turn-guard-right {:guard-position {:direction :up :row 6 :col 4}})))
      (is (= {:direction :down :row 6 :col 4}
             (gm/turn-guard-right {:guard-position {:direction :right :row 6 :col 4}})))
      (is (= {:direction :left :row 6 :col 4}
             (gm/turn-guard-right {:guard-position {:direction :down :row 6 :col 4}})))))
  (testing "guard-path"
    (testing "trace guard path in map until they exit"
      (is (= [{:direction :up :row 1 :col 0}
              {:direction :up :row 0 :col 0}]
             (gm/guard-path {:guard-map {:map ["."
                                               "^"]
                                         :width 1 :height 2}})))
      (is (= [{:direction :up :row 2 :col 0}
              {:direction :right :row 1 :col 0}
              {:direction :right :row 1 :col 1}]
             (gm/guard-path {:guard-map {:map ["#."
                                               ".."
                                               "^."]
                                         :width 2 :height 3}})))
      (is (= {:map ["....#....."
                    "....>>>>v#"
                    "....^...v."
                    "..#.^...v."
                    "..>>>>v#v."
                    "..^.^.v.v."
                    ".#^<<<v<<."
                    ".>>>>>>v#."
                    "#^<<<<<v.."
                    "......#v.."]
              :width 10
              :height 10}
             (let [guard-map (gm/parse-guard-map-file {:file-name "data/day06/test.txt"})]
               (gm/visited-map {:guard-map guard-map}))))))
  (testing "possible-loop-obstructions"
    (testing "find all possible obstructions leading to a loop on guard path"
      (is (= [{:row 6 :col 3}
              {:row 7 :col 6}
              {:row 7 :col 7}
              {:row 8 :col 1}
              {:row 8 :col 3}
              {:row 9 :col 7}]
             (let [guard-map (gm/parse-guard-map-file {:file-name "data/day06/test.txt"})
                   guard-path (gm/guard-path {:guard-map guard-map})]
               (gm/possible-loop-obstructions {:guard-map guard-map
                                               :guard-path guard-path})))))))

File: js/day06/guardMap.test.js
require('../init');
const {
  parseGuardMapFile,
  getInitialPosition,
  guardIsExiting,
  advanceGuard,
  turnGuardRight,
  traceGuardPath,
  findAllPossibleLoopObstructions,
  visitedMap,
} = require('./guardMap');

describe('day06', () => {
  it('parseGuardMapFile', () => {
    const { guardMap } = parseGuardMapFile({ fileName: 'data/day06/test.txt' });
    expect({ guardMap }).to.eql({
      guardMap: {
        map: [
          '....#.....',
          '.........#',
          '..........',
          '..#.......',
          '.......#..',
          '..........',
          '.#..^.....',
          '........#.',
          '#.........',
          '......#...',
        ],
        height: 10,
        width: 10,
      },
    });
  });

  describe('getInitialPosition', () => {
    it('extracts initial position from map, denoted "^"', () => {
      const { guardMap } = parseGuardMapFile({
        fileName: 'data/day06/test.txt',
      });
      expect(
        getInitialPosition({
          guardMap,
        })
      ).to.eql({
        direction: 'up',
        row: 6,
        col: 4,
      });
    });
  });

  describe('guardIsExiting', () => {
    it('checks if advancing the guard means they exit the map', () => {
      const guardMap = { width: 10, height: 10 };
      expect(
        guardIsExiting({
          guardMap,
          guardPosition: { direction: 'up', row: 4, col: 4 },
        })
      ).to.eql(false);
      expect(
        guardIsExiting({
          guardMap,
          guardPosition: { direction: 'up', row: 0, col: 4 },
        })
      ).to.eql(true, 'up');
      expect(
        guardIsExiting({
          guardMap,
          guardPosition: { direction: 'down', row: 9, col: 4 },
        })
      ).to.eql(true, 'down');
      expect(
        guardIsExiting({
          guardMap,
          guardPosition: { direction: 'left', row: 4, col: 0 },
        })
      ).to.eql(true, 'left');
      expect(
        guardIsExiting({
          guardMap,
          guardPosition: { direction: 'right', row: 4, col: 9 },
        })
      ).to.eql(true, 'right');
    });
  });

  describe('advanceGuard', () => {
    it('advance guard in their current direction', () => {
      expect(
        advanceGuard({
          guardPosition: { direction: 'up', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'up', row: 3, col: 4 });
      expect(
        advanceGuard({
          guardPosition: { direction: 'down', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'down', row: 5, col: 4 });
      expect(
        advanceGuard({
          guardPosition: { direction: 'left', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'left', row: 4, col: 3 });
      expect(
        advanceGuard({
          guardPosition: { direction: 'right', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'right', row: 4, col: 5 });
    });
  });

  describe('turnGuardRight', () => {
    it('change guard direction to turn right', () => {
      expect(
        turnGuardRight({
          guardPosition: { direction: 'up', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'right', row: 4, col: 4 });
      expect(
        turnGuardRight({
          guardPosition: { direction: 'right', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'down', row: 4, col: 4 });
      expect(
        turnGuardRight({
          guardPosition: { direction: 'down', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'left', row: 4, col: 4 });
      expect(
        turnGuardRight({
          guardPosition: { direction: 'left', row: 4, col: 4 },
        })
      ).to.eql({ direction: 'up', row: 4, col: 4 });
    });
  });

  describe('traceGuardPath', () => {
    it('trace guard in map until they exit', () => {
      expect(
        traceGuardPath({
          guardMap: {
            map: ['.', '^'],
            width: 1,
            height: 2,
          },
        })
      ).to.eql([
        { direction: 'up', row: 1, col: 0 },
        { direction: 'up', row: 0, col: 0 },
      ]);
    });

    it('turn guard to the right when they encounter an obstruction', () => {
      expect(
        traceGuardPath({
          guardMap: {
            map: ['#.', '..', '^.'],
            width: 2,
            height: 3,
          },
        })
      ).to.eql([
        { direction: 'up', row: 2, col: 0 },
        { direction: 'up', row: 1, col: 0 },
        { direction: 'right', row: 1, col: 1 },
      ]);
    });

    it('trace test map', () => {
      const { guardMap } = parseGuardMapFile({
        fileName: 'data/day06/test.txt',
      });
      const guardPath = traceGuardPath({ guardMap });
      expect(visitedMap({ guardMap, guardPath })).to.eql([
        '....#.....',
        '....^>>>>#',
        '....^...v.',
        '..#.^...v.',
        '..^>>>>#v.',
        '..^.^.v.v.',
        '.#<<<<v<v.',
        '.^>>>>>>#.',
        '#<<<<<vv..',
        '......#v..',
      ]);
    });
  });

  describe('findAllPossibleLoopObstructions', () => {
    it('checks along guard path for any possible obstruction leading them into a loop', () => {
      const { guardMap } = parseGuardMapFile({
        fileName: 'data/day06/test.txt',
      });
      const guardPath = traceGuardPath({ guardMap });
      expect(findAllPossibleLoopObstructions({ guardMap, guardPath })).to.eql([
        { row: 6, col: 3 },
        { row: 7, col: 6 },
        { row: 7, col: 7 },
        { row: 8, col: 1 },
        { row: 8, col: 3 },
        { row: 9, col: 7 },
      ]);
    });
  });
});
